nca/FWRules/MinimizeCsFWRulesOpt.py

```diff
@@ -0,0 +1,644 @@
+        self._compute_basic_namespace_grouping()
+
+        # add all fw-rules:
+        self._add_all_fw_rules()
+
+    def _compute_basic_namespace_grouping(self):
+        """
+        computation of peer sets with possible grouping by namespaces.
+        Results are at: ns_set_pairs, base_elem_pairs, peer_props_without_ns_expr
+        :return: None
+        """
+        self._compute_covered_peer_props()
+        # only Pod elements have namespaces (skipping IpBlocks and HostEPs)
+        all_src_ns_set = set(src.namespace for src in self.covered_peer_props.project_on_one_dimension("src_peers")
+                             if isinstance(src, Pod))
+        all_dst_ns_set = set(dst.namespace for dst in self.covered_peer_props.project_on_one_dimension("dst_peers")
+                             if isinstance(dst, Pod))
+        self._compute_full_ns_grouping(all_src_ns_set, all_dst_ns_set)
+        src_peers_without_ns = PeerSet(set(src for src in self.peer_props.project_on_one_dimension("src_peers")
+                                           if isinstance(src, (IpBlock, HostEP, DNSEntry))))
+        dst_peers_without_ns = PeerSet(set(dst for dst in self.peer_props.project_on_one_dimension("dst_peers")
+                                           if isinstance(dst, (IpBlock, HostEP, DNSEntry))))
+        props_with_elems_without_ns = \
+            ConnectivityProperties.make_conn_props_from_dict({"src_peers": src_peers_without_ns}) |\
+            ConnectivityProperties.make_conn_props_from_dict({"dst_peers": dst_peers_without_ns})
+        self.peer_props_without_ns_expr |= props_with_elems_without_ns & self.peer_props
+        # compute pairs with src as pod/ip-block and dest as namespace
+        self._compute_partial_ns_grouping(all_dst_ns_set, False)
+        # compute pairs with src as pod/ip-block namespace dest as pod
+        if self.peer_props_without_ns_expr:
+            self._compute_partial_ns_grouping(all_src_ns_set, True)
+        if self.peer_props_without_ns_expr:
+            self._compute_full_ipblock_and_dns_grouping(False)
+        if self.peer_props_without_ns_expr:
+            self._compute_full_ipblock_and_dns_grouping(True)
+
+    def _compute_covered_peer_props(self):
+        """
+        compute the union (set) of all peer pairs for which communication is allowed in current connection-set (but
+        not necessarily only limited to current connection set)
+        :return: None
+        """
+        covered_peer_props = self.peer_props | self.peer_props_in_containing_connections
+        all_peers_set = self.peer_props.get_all_peers()
+        for pod in all_peers_set:
+            if isinstance(pod, ClusterEP):
+                covered_peer_props |= ConnectivityProperties.make_conn_props_from_dict({"src_peers": PeerSet({pod}),
+                                                                                        "dst_peers": PeerSet({pod})})
```

Comment:
this captures only pairs where both src and dst are `ClusterEP` ?
in the original version it was pairs where at lease one of these elements was  `ClusterEP`
```
        all_pods_set = set(src for (src, dst) in self.peer_pairs if isinstance(src, ClusterEP)) | \
            set(dst for (src, dst) in self.peer_pairs if isinstance(dst, ClusterEP))
        for pod in all_pods_set:
            covered_peer_pairs_union |= {(pod, pod)}
```

Reply:
It calculates all `ClusterEP peers` that appear in connections. The original calculation took them from `peer_pairs` (taking `src` if `src` is `ClusterEP` or `dst` if `dst` is `ClusterEP`).
The optimized calculation first computes `all_peers_set` by taking the union of projection of connections to `src_peers` and to `dst_peers`, and then takes from them those peers that are `ClusterEP`.
Thus, the both original and optimized calculations are equivalent.

Reply:
but in this version both src and dst are `ClusterEP` , and in the original version only one of src or dst?